test/envoye2e/stats/stats_xds_test.go

```diff
@@ -0,0 +1,233 @@
+			&driver.Envoy{Bootstrap: params.LoadTestData("testdata/bootstrap/client.yaml.tmpl")},
+			&driver.Sleep{1 * time.Second},
+			&driver.Repeat{N: 10, Step: &driver.Get{ports.ClientToServerProxyPort, "hello, world!"}},
+			&driver.Stats{ports.ClientAdminPort, map[string]driver.StatMatcher{
+				"istio_requests_total": &driver.ExactStat{"testdata/metric/client_request_total.yaml.tmpl"},
+			}},
+			&driver.Stats{ports.ServerAdminPort, map[string]driver.StatMatcher{
+				"istio_requests_total": &driver.ExactStat{"testdata/metric/server_request_total.yaml.tmpl"},
+			}},
+		},
+	}).Run(params); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestStatsParallel(t *testing.T) {
+	ports := env.NewPorts(env.StatsParallel)
+	params := &driver.Params{
+		Vars: map[string]string{
+			"ClientPort":   fmt.Sprintf("%d", ports.ClientToServerProxyPort),
+			"BackendPort":  fmt.Sprintf("%d", ports.BackendPort),
+			"ClientAdmin":  fmt.Sprintf("%d", ports.ClientAdminPort),
+			"ServerAdmin":  fmt.Sprintf("%d", ports.ServerAdminPort),
+			"ServerPort":   fmt.Sprintf("%d", ports.ProxyToServerProxyPort),
+			"RequestCount": "1",
+		},
+		XDS: int(ports.XDSPort),
+	}
+	params.Vars["ClientMetadata"] = params.LoadTestData("testdata/client_node_metadata.json.tmpl")
+	params.Vars["ServerMetadata"] = params.LoadTestData("testdata/server_node_metadata.json.tmpl")
+	params.Vars["StatsConfig"] = params.LoadTestData("testdata/bootstrap/stats.yaml.tmpl")
+	client_request_total := &dto.MetricFamily{}
+	server_request_total := &dto.MetricFamily{}
+	params.LoadTestProto("testdata/metric/client_request_total.yaml.tmpl", client_request_total)
+	params.LoadTestProto("testdata/metric/server_request_total.yaml.tmpl", server_request_total)
+
+	if err := (&driver.Scenario{
+		[]driver.Step{
+			&driver.XDS{},
+			&driver.Update{Node: "client", Version: "0", Listeners: []string{StatsClientHTTPListener}},
+			&driver.Update{Node: "server", Version: "0", Listeners: []string{StatsServerHTTPListener}},
+			&driver.Envoy{Bootstrap: params.LoadTestData("testdata/bootstrap/server.yaml.tmpl")},
+			&driver.Envoy{Bootstrap: params.LoadTestData("testdata/bootstrap/client.yaml.tmpl")},
+			&driver.Sleep{1 * time.Second},
+			&driver.Get{ports.ClientToServerProxyPort, "hello, world!"},
+			&driver.Fork{
+				Fore: &driver.Scenario{
+					[]driver.Step{
+						&driver.Sleep{1 * time.Second},
+						&driver.Repeat{
+							Duration: 9 * time.Second,
+							Step:     &driver.Get{ports.ClientToServerProxyPort, "hello, world!"},
+						},
+						capture{},
+					},
+				},
+				Back: &driver.Repeat{
+					Duration: 10 * time.Second,
+					Step: &driver.Scenario{
+						[]driver.Step{
+							&driver.Update{Node: "client", Version: "{{.N}}", Listeners: []string{StatsClientHTTPListener}},
+							&driver.Update{Node: "server", Version: "{{.N}}", Listeners: []string{StatsServerHTTPListener}},
+							// may need short delay so we don't eat all the CPU
+							&driver.Sleep{10 * time.Millisecond},
```

Comment:
Maybe make it sleep longer so that stats parsing could have more cpu cycle?

Reply:
Stats parsing happens after this loop finishes, so it should not have an effect I think.

Reply:
Oh, so stats parsing only happens when /stats/prometheus is hit? Then it is that stats increment is not efficient enough and cause data path to time out because prow pod lacks CPU?